### ISSUE TITLE
EntityDecoder's orElse delegates to the right decoder irrespective of specified MediaRange

### ISSUE THREAD

**Issue by jedesah** [http4s/http4s#436]
The current implementation of `orElse` checks the `MediaType` of the incoming `Request`, passes it along to the first `EntityDecoder` if it satisfies the `MediaType` or else passes it along to the other `EntityDecoder` no matter what.

I think it would be better if the implementation also checked whether the accepts of the second `EntityDecoder` was also appropriate and handles the failure case itself in the event that the second also does not support the `MediaType`. This seems better then attempting to parse the body and probably error out with some bogus error message.

I am happy to make this change since I have my mind pretty much wrapped around it. I did want to post this issue first though to get confirmation that this would be considered a positive change because the proper fixe involves changing a few types. Ideally we want the error from an `orElse` Decoder to propagate to a `UnsupportedMediaType("Media-Type is missing")` and that involves changing the type of `DecodeResult` in order for the implementation of `Request.decode` to identify that kind of failure and generate the appropriate `UnsupportedMediaType` error Response.

**Comment by bryce-anderson** [comment]
I think I can get behind not using the second decoder arbitrarily, but we need to work out what happens when no media type is specified. I would also like to know what you would plan for `DecodeResult`. What would the new `DecodeResult` look like? Does it remain an `EitherT[Task, ParseFailure, R]`, and `ParseFailure` gets expanded into a family of types or something else?

**Comment by jedesah** [comment]
So first of all, just to clarify my understanding that if one does not use `orElse`, the `EntityDecoder` will attempt to decode the `Message` irrespective. That is, an `EntityDecoder` does not call the `matchesMediaType` all by itself. I think that behavior is probably fine.

I can think of two possibilities for how to handle the case where the `Message` does not specify a `Media-Type` and `orElse` is being used.

The first one is to say that one needs to add a catch all, i.e `*/*` at the end of the `orElse` chain if they explicitly want to handle the case of a missing `Media-Type`. That would require the implementation to change in order for `*/*` to accept `None`. Currently, we just return false if we see `None` in `matchesMediaType`.

The second one is to change the implementation of `OrDec` to only handle failure itself if they are explicitly different but, to send the `Message` to the "last" `Decoder` like it does currently if no media-type is specified in the `Message`.

I think I would prefer the first scenario, but I may be biased because that would best for Quasar because we want to force the `Media-Type` to be specified in some of our REST API endpoints. I do think that both approaches are probably better than the current behavior so I am happy to implement the second approach if we prefer how it is more lenient towards a loosely specified `Message`.

I think if `DecodeResult` becomes a `EitherT[Task, DecodeFailure, R]` where `DecodeFailure` is a either a `ParseFailure` or a `MediaTypeMismatch` or a `MediaTypeMissing` (do we want to make a distinction?), that should do the trick.

**Comment by bryce-anderson** [comment]
Is there value in being able to specify whether the decoder must be explicitly satisfied or if it will yolo it at the end if no media-type was found? I can see both situations being useful so I suspect there is.

**Comment by jedesah** [comment]
Sure, you mean like a boolean parameter for `orElse`? If so it should
probably have a default value in order to make it easy to chain them
together (can one use infix with a two parameter function that has a
default?). If it does, then I guess the question becomes what should be the
default value? :smile:

**Comment by bryce-anderson** [comment]
Maybe I'm going to backtrack a little bit on the boolean parameter on `OrDec`: the same question could be asked of a single decoder. Maybe the right thing to do is add such the flag to the base decoders and that can bubble up to the `OrDec`. For example
```scala
val yolo = EntityDecoder[ByteVector].yolo(true) // I suspect this one would be yolo by default
val strict1 = EntityDecoder[JSON]
val strict2 = EntityDecoder[String]
(strict1 orElse strict2).isYolo == false // neither are yolo decoders...
```
The naming of methods is only for example...

**Comment by bryce-anderson** [comment]
@jedesah, do you want to try and get this in before the next release? I'd like to get the next release out soon and I expect this to be a binary incompatible change.

**Comment by jedesah** [comment]
Yes, that would be cool. let me have a look at it tonight or tmr.

**Comment by bryce-anderson** [comment]
@jedesah, awesome, thanks!

**Comment by bryce-anderson** [comment]
Closed via #445.
